### ISSUE TITLE
[BUG] Kva certificate needs to be deleted if Kva Certificate expired after 60 days

### ISSUE THREAD

**Issue by raghavendra-nataraj** [Azure/aksArc#146]
**Describe the bug**
KVA certificate expires after 60 day if no upgrade is performed. 

**Expected behavior**
Update-AksHci would throw the below error. Any command involving kvactl will throw the below error. 
`Error: failed to get new provider: failed to create azurestackhci session: Certificate has expired: Expired`

**Environment:**
- AKS-HCI Version (Any AksHci version on or before 1.0.4.10928)

**Solution**
Delete the expired certificate file at the below location and try Update again
$env:UserProfile\.wssd\kvactl\cloudconfig

**Comment by madhanrm** [comment]
Should this be deleted on all nodes?

**Comment by raghavendra-nataraj** [comment]
Only on the node seeing the certificate expiry issue

**Comment by gianniskt** [comment]
The file `$env:UserProfile.wssd\kvactl\cloudconfig`  contents are like this format:

![image](https://user-images.githubusercontent.com/38279583/141418201-533f159b-4011-4d0c-af8b-4a15edfd827f.png)

Do you mean that we have to delete the block `<SOME-CERTIFICATE-DATA>` ? I tried this solution, and the above error is shown:

![image](https://user-images.githubusercontent.com/38279583/141418369-c0fc7962-cd27-4d4c-9bc0-b4500c3bce19.png)

There is no other certificate file at this location.

**Comment by raghavendra-nataraj** [comment]
Please can you delete the entire file and try again.

**Comment by gianniskt** [comment]
Ok so I Deleted the "cloudconfig" file. After `Update-AksHci` a new "cloudconfig" file is generated, and the below error is shown:

![image](https://user-images.githubusercontent.com/38279583/141425296-c8e9db0f-ef11-4865-ab51-9105df1ab536.png)

**Comment by raghavendra-nataraj** [comment]
Looking at the other error in the bug you posted it looks like the KMS is down as it crossed the 60 day expiry. Can you do verify that by running the below command `kubectl --kubeconfig (Get-KvaConfig).Kubeconfig get secrets -A`. If the command does not work It could be that the KMS is down. Can you try this [workaround](https://github.com/Azure/aks-hci/issues/127)

**Comment by gianniskt** [comment]
Yes I have tried this workaround, but on step 3, as you can see at the screenshot, the container `kms-plugin` has status `Exited`

![image](https://user-images.githubusercontent.com/38279583/141616056-32ede7b9-4ab2-4776-b11e-fac4f2623d04.png)

So I cannot start the container. Also many other containers are in same status. This happened after 60 days.
I am thinking of applying the command `Restart-AksHci` , but I think this will also remove the workload cluster, which is working fine, am I correct?

**Comment by raghavendra-nataraj** [comment]
Yes, It would remove the workload cluster. We can try to understand what is causing the KMS plugin to crash by looking at the logs. 
`$ ssh -i (Get-AksHciConfig).Moc.sshPrivateKey clouduser@<ip> "sudo docker container ls | grep 'kms'"`
The output should have the below fields
`CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES`
The output should look something like this
`4169c10c9712 f111078dbbe1 "/kms-plugin" 22 minutes ago Up 22 minutes k8s_kms-plugin_kms-plugin-moc-lll6bm1plaa_kube-system_d4544572743e024431874e6ba7a9203b_1`
We can get the logs with 
`$ ssh -i (Get-AksHciConfig).Moc.sshPrivateKey clouduser@<ip> "sudo docker container logs <Container ID>"`

**Comment by gianniskt** [comment]
Hello,

Because the container runtime is "containerd" & not "Docker", the right command is: `crictl stop <Container ID> && crictl start <Container ID>`
The issue now is resolved.
Now the problem is that it cannot communicate with the workload cluster:

![get-akshciupdates](https://user-images.githubusercontent.com/38279583/141701004-1a6e3179-650e-4b50-9204-305b10a84e53.JPG)

Also when trying to run `Update-AksHci` the error message is:

![image](https://user-images.githubusercontent.com/38279583/141701033-5645582e-6362-4306-872d-efda5a5c6a10.png)

Should I open new issue? Or you can suggest me what to look for troubleshooting?

**Comment by raghavendra-nataraj** [comment]
So is the management cluster api server accessible ? Can you try running this command 
`kubectl --kubeconfig (Get-KvaConfig).Kubeconfig get nodes`

**Comment by gianniskt** [comment]
Yes the management cluster API Server is accessible. I attach the commands that have run for both management & workload clusters:

![image](https://user-images.githubusercontent.com/38279583/141940664-687ec6ec-a1b5-4802-9d9d-6ac2d2bef780.png)

**Comment by Elektronenvolt** [comment]
Hi @gianniskt 

I was like you in the mud with another issue [148](https://github.com/Azure/aks-hci/issues/148) 
With a bit help from Zach I made it with following steps:

`kubectl get machines -A --kubeconfig (Get-AksHciConfig).Kva.kubeconfig` showed me the old management cluster VM being stuck at deleting with kube-apiserver and etcd in a crash loop.
![image](https://user-images.githubusercontent.com/6702311/142024300-aff30637-0e4c-4048-a595-7b61bbc36681.png)

I did a manual shutdown of the old VM in Hyper-V -> After that it had been deleted automatically, but I had the new management cluster VM and `Get-AksHciVersion` showed me the old version 1.0.4 and `Update-AksHci`  the same error message 'update in progress'

Zach asked me to set the state manually to "UpdateFailed" to overcome this exception - I've modified AksHci.psm to
![image](https://user-images.githubusercontent.com/6702311/142025697-75c370aa-9f09-4e87-ad6b-574a09f2c4c5.png)

Running `Update-AksHci`  again and it successfully finished the update (without creating a new management cluster VM)
@raghavendra-nataraj  may this helps here as well to finalize the update?

**Comment by gianniskt** [comment]
Hello @Elektronenvolt

The error for my case is different:

![image](https://user-images.githubusercontent.com/38279583/142054214-d585a43d-cc71-4d16-afeb-fe2570cdfad8.png)

As I understand, it is trying to provision a new management control-plane VM, and it is stuck in provisioning.
I cannot see at the HyperVisor any new VM that is being created, or any job related to this VM Name.

I also tried your workaround but the result of `Update-AksHci` is the same `Update is already in progress`.

**Comment by Elektronenvolt** [comment]
@gianniskt I see, I thought you're hanging on tearing down the old VM.

**Comment by raghavendra-nataraj** [comment]
Thanks @gianniskt, we may need logs to debug this issue further. We are working on the process for sharing logs. @Elektronenvolt thanks for steps. It looks like issue here is the vm is not getting created. So it may not be the right workaround here.

**Comment by gianniskt** [comment]
Do you need log zip file extracted from command `Get-AksHciLogs` ?
The file is almost 2.5 GB. Is there maybe any specific folder of the Log File that you may need?

**Comment by raghavendra-nataraj** [comment]
Yes, I would need that. We are looking at the process to get that. Meanwhile can you check for caph logs in folder "\Kva\clusterlogs\caph-system\caph-controller-manager-*\logs"

**Comment by madhanrm** [comment]
kubectl get event --kubeconfig (Get-AksHciConfig).Kva.kubeconfig -A

**Comment by gianniskt** [comment]
[aks_hci_event_logs.txt](https://github.com/Azure/aks-hci/files/7560661/aks_hci_event_logs.txt)

You can find attached the output of the above command.

**Comment by madhanrm** [comment]
does this image exist -> C:\ClusterStorage\Volume1\AksHCI\ImageStore\dfa6f14521777d5\Linux_k8s_1-21-2.vhdx

**Comment by madhanrm** [comment]
Can you get the output of the cmdlet
Get-MocGalleryImage -Location MocLocation

**Comment by gianniskt** [comment]
The output of the command is:

![image](https://user-images.githubusercontent.com/38279583/142372940-25f62730-a5df-43b9-bda3-982bbdce97dc.png)

However I cannot see the image file at the path `C:\ClusterStorage\Volume1\AksHCI\ImageStore\dfa6f14521777d5\Linux_k8s_1-21-2.vhdx` .

This is the image that exists at the folder:

![image](https://user-images.githubusercontent.com/38279583/142373248-d8f83e51-bdf3-44bd-8fbe-8a3f1a93866a.png)

**Comment by madhanrm** [comment]
Not sure how that image went missing, but lets try to restore it, to let the upgrade continue.
Can you check if you find Linux_k8s_1-21-2.vhdx is in the parent dir?

If so, can you copy it here (C:\ClusterStorage\Volume1\AksHCI\ImageStore\dfa6f14521777d5\Linux_k8s_1-21-2.vhdx)

**Comment by gianniskt** [comment]
Yes, that was my first action after checking this error, but I cannot find this file anywhere.

**Comment by madhanrm** [comment]
Okay, lets try the below then

```
Remove-MocGalleryImage -location MocLocation  -Name Linux_k8s_1-21-2
ipmo kva.psm1
Add-KvaGalleryImage -kubernetesVersion v1.21.2
```

**Comment by madhanrm** [comment]
If this succeeds, you should see the upgrade progressing and new vms getting created.
Upgrade should be done in 15 minutes

**Comment by madhanrm** [comment]
If you are stuck in Upgrade is in Progress, do take a look at the below issue to see how to fix it and try again.

https://github.com/Azure/aks-hci/issues/152 

But do this only if there isnt any upgrade window that is in progress via PS or WAC.

**Comment by gianniskt** [comment]
Ok these commands worked, and the new management cluster VM has been provisioned:

![image](https://user-images.githubusercontent.com/38279583/142390251-e464b062-c7dc-4956-8ed9-bf592f3e4640.png)

However, the old management VM `aks-interworks-cluster1-control-plane-0` is still there, and the following error about this node is displayed:

![image](https://user-images.githubusercontent.com/38279583/142390477-8ef07e25-90cd-4cd1-a6ca-2277ae35069e.png)

Should the old management cluster VM be deleted, or look up other logs?

**Comment by gianniskt** [comment]
Update...
The old management cluster VM has been removed automatically.

![image](https://user-images.githubusercontent.com/38279583/142399075-594d16a5-72ef-4570-876f-daac030991ec.png)

However the same error is displayed:

![image](https://user-images.githubusercontent.com/38279583/142399280-547a5dad-d799-4c9c-8e01-1cfa25bbf85c.png)

The new log file is attached:

[aks_hci_new_event_logs.txt](https://github.com/Azure/aks-hci/files/7562061/aks_hci_new_event_logs.txt)

**Comment by madhanrm** [comment]
Can you check if below command works? Share the output, only if the command fails with an error.

```
kubectl get secrets --kubeconfig (Get-AksHciConfig).Kva.kubeconfig
```

**Comment by madhanrm** [comment]
@raghavendra-nataraj , do you know what this error means. Appliance is reachable, but these cmds are failing

**Comment by gianniskt** [comment]
@madhanrm Yes the command `kubectl get secrets --kubeconfig (Get-AksHciConfig).Kva.kubeconfig` is working.

**Comment by madhanrm** [comment]
trying to find out what the next steps are.

**Comment by gianniskt** [comment]
Also to provide more information, this is the output of command `Get-AksHciBillingStatus`

![image](https://user-images.githubusercontent.com/38279583/142619933-2d45b996-a4ec-4954-a43e-a8fecad3bffe.png)

However, the 2 clusters (Management & Workload) are shown connected at Azure-Arc.

Furthermore, you can see the structure at the AksHci folder:

![image](https://user-images.githubusercontent.com/38279583/142620405-db5a2d4b-8993-4f06-beb5-657b28802d8c.png)

The `1.0.3.10901` is the current version of AksHci, and the `1.0.4.10928` is the next version that `Update-AksHci` was trying to apply.

**Comment by baziwane** [comment]
@gianniskt to share the logs, can you use your Azure Subscription to create a support request: https://docs.microsoft.com/en-us/azure/azure-portal/supportability/how-to-create-azure-support-request. Make sure you choose Azure Kubernetes Service on Azure Stack HCI in the service type so that your issue is correctly routed.
In Step 3: Additional details you'll have option to upload a file.

**Comment by gianniskt** [comment]
I cannot create an Azure Technical Support Request as I can see that we have the 'Basic Plan' for our Subscription and the 'Developer Plan' is needed for this request.

I will check if we can upgrade to 'Developer Plan' but either way I posted a question at Q&A:
https://docs.microsoft.com/en-us/answers/questions/635831/cannot-update-akshci-cluster-even-all-nodes-are-up.html

**Comment by baziwane** [comment]
@gianniskt could you share the support ticket # if you still have it by any chance

**Comment by gianniskt** [comment]
@baziwane Support request number: 2111220050001803

**Comment by gianniskt** [comment]
We decided to uninstall the AKS-HCI, and install it again, so the issue will be closed.

**Comment by mattbriggs** [comment]
@mattbriggs 

#in-progress
